rewrite-test/src/main/kotlin/org/openrewrite/java/typegeneration/TypeGenerationTest.kt

```diff
@@ -0,0 +1,512 @@
+/*
+ * Copyright 2020 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.typegeneration
+
+import com.fasterxml.jackson.annotation.JsonIdentityInfo
+import org.assertj.core.api.Assertions.assertThat
+import org.intellij.lang.annotations.Language
+import org.junit.jupiter.api.Disabled
+import org.junit.jupiter.api.Test
+import org.objectweb.asm.*
+import org.objectweb.asm.util.ASMifier
+import org.objectweb.asm.util.TraceClassVisitor
+import org.openrewrite.InMemoryExecutionContext
+import org.openrewrite.Parser
+import org.openrewrite.java.JavaParser
+import org.openrewrite.java.tree.J
+import java.io.ByteArrayOutputStream
+import java.io.PrintStream
+import java.io.PrintWriter
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.util.*
+
+
+interface TypeGenerationTest {
+
+    // This test provides a generic way to assert that a java source (as a string or as a path) can first be
+    // compiled using the runtime classpath and then it will extract the types from the resulting AST and attempt
+    // to compile the source again using just the generated byte code.
+    //
+    // The two variants of assertCompilationWithByteCode only require a source, but allows for the parser, and two
+    // method callbacks to be overridden to assert test conditions.
+    //
+    // The system.out is captured and passed to the compiler out handler (and the default is to swallow that output)
+    // checkCompilationUnit can be use used to verify type attribution on the compilation unit after the second parse.
+
+    val rewriteRoot: Path
+        get () {
+            val normalize = Paths.get(TypeGenerationTest::class.java.getResource("./")!!.toURI())
+                .resolve("../../../../../../../../").normalize()
+            return normalize
+        }
+
+    val parser: JavaParser
+        get() = JavaParser.fromJavaVersion()
+            .logCompilationWarningsAndErrors(true)
+            .build()
+
+    //Default is to swallow compiler output, this can be overridden for any given test.
+    val defaultCompilerOutputHandler: (String) -> Unit
+        get() = {}
+
+    @Test
+    fun testCompilationUsingExtractedTypes() {
+        assertCompilationWithByteCode(
```

Comment:
It would be stylistically consistent with our other tests to write these tests as single-expression methods.
e.g.: `fun testCompilationUsingExtractedTypes() = assertCompilationWithByteCode(...)`